File: tools/observation_vector/mean_rev_z_variant_comparison.cpp
// mean_rev_z_variant_comparison.cpp -- empirical additive-vs-replace decision
// tool for mean_rev_z/fast_mean_rev_z, per the sibling's revised verdict
// (SCRATCHPAD.md, 2026-08-28): correlation alone is not the deciding test:
// feed each variant through Scoring.cpp:305's actual gate condition
// (score > 2.0f) and measure forward-return/hit-rate on the resulting
// signals, independently per variant, on real MES history.
//
// Uses the REAL, unmodified include/ActivityClockMeanReversion.h
// (ActivityClockMeanRevZ) for the activity-clock side -- same "exact port,
// drive with real data" precedent as tools/observation_vector/fractal_dim_threshold_migration.cpp.
// The time-bar side replicates CalculateMeanReversionSpeed's exact math
// (StudyHelperFunctions.cpp:3247) as a faithful reference port, the same
// technique tests/cpp/test_dfa_hurst_exponent.cpp's BruteForceHurst uses --
// the original function is ACSIL-embedded (takes SCStudyInterfaceRef) and
// cannot be #included standalone.
//
// Known, explicitly-flagged approximation: real Sierra Chart tick-level
// signed order flow is not available offline; this tool builds imbalance
// bars from mes_continuous_ticks.parquet's 1-second aggregated
// bid_volume/ask_volume columns instead (confirmed 1-second bars, not
// per-trade ticks, earlier this project). Each 1-second row is fed to
// ImbalanceBarEngine as its own atomic "tick" (a fresh barIndex per row
// forces its delta computation to use that row's own signed volume
// directly) -- this preserves real net per-second order-flow imbalance and
// its cumulative accumulation across seconds, the core AFML imbalance-bar
// mechanism, but loses true intra-second granularity. The engine's
// production imbalance threshold (50.0f) was calibrated for real
// tick-level cumulative signed volume between chart bars, not 1-second
// aggregates -- this tool empirically picks its own threshold (see
// kDefaultImbalanceThreshold below) to produce a reasonable bar-formation
// rate on this coarser data, not the live production value.
//
// I/O is raw binary (contiguous arrays, no text parsing) -- fed by
// tools/observation_vector/mean_rev_z_variant_comparison.py via polars/numpy, not pandas/CSV
// (38.5M-row CSV round-trips were the wrong tool for this data volume).
// Format: [int64 count][int64 timestamp_us x count][float32 field2 x count]
// [float32 field3 x count (ticks file only)][float32 field4 x count (ticks only)].
//
// Build: g++ -O2 -std=c++17 -Iinclude tools/observation_vector/mean_rev_z_variant_comparison.cpp -o mean_rev_z_variant_comparison
// Usage: ./mean_rev_z_variant_comparison ticks_1s.bin bars_15m.bin signals_out.csv [imbalance_threshold]
//   ticks_1s.bin: count, timestamp_us[], close[], bid_volume[], ask_volume[]
//   bars_15m.bin: count, timestamp_us[], close[]
//   signals_out.csv: one row per gate-crossing signal (either variant), for
//                     the companion Python script to compute forward returns
//                     against the full 1-second price series and report
//                     hit-rate/forward-return statistics per variant.
//   imbalance_threshold: optional, defaults to kDefaultImbalanceThreshold below --
//                        pass a candidate value to calibrate the activity
//                        clock's bar-formation rate before trusting results.

#include "ActivityClockMeanReversion.h"
#include "ImbalanceBarEngine.h"

#include <algorithm>
#include <array>
#include <cmath>
#include <cstdint>
#include <cstdlib>
#include <deque>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

namespace {

constexpr int kReturnBufferN = 100;         // matches ContextManager.cpp's shared 100-return fetch
constexpr float kGateThreshold = 2.0f;      // Scoring.cpp:305's real threshold
constexpr float kDefaultImbalanceThreshold = 15.0f; // fallback if not passed on the command line

struct Tick {
    long long timestamp_us;
    float close;
    float bid_volume;
    float ask_volume;
};

struct Bar15m {
    long long timestamp_us;
    float close;
};

template <typename T>
std::vector<T> ReadArray(std::ifstream& in, std::int64_t count) {
    std::vector<T> out(static_cast<std::size_t>(count));
    in.read(reinterpret_cast<char*>(out.data()), static_cast<std::streamsize>(count * static_cast<std::int64_t>(sizeof(T))));
    return out;
}

std::vector<Tick> LoadTicks(const std::string& path) {
    std::ifstream in(path, std::ios::binary);
    std::int64_t count = 0;
    in.read(reinterpret_cast<char*>(&count), sizeof(count));
    const auto ts = ReadArray<std::int64_t>(in, count);
    const auto close = ReadArray<float>(in, count);
    const auto bid = ReadArray<float>(in, count);
    const auto ask = ReadArray<float>(in, count);

    std::vector<Tick> out(static_cast<std::size_t>(count));
    for (std::int64_t i = 0; i < count; ++i) {
        out[static_cast<std::size_t>(i)] = {ts[static_cast<std::size_t>(i)], close[static_cast<std::size_t>(i)],
                                             bid[static_cast<std::size_t>(i)], ask[static_cast<std::size_t>(i)]};
    }
    return out;
}

std::vector<Bar15m> LoadBars15m(const std::string& path) {
    std::ifstream in(path, std::ios::binary);
    std::int64_t count = 0;
    in.read(reinterpret_cast<char*>(&count), sizeof(count));
    const auto ts = ReadArray<std::int64_t>(in, count);
    const auto close = ReadArray<float>(in, count);

    std::vector<Bar15m> out(static_cast<std::size_t>(count));
    for (std::int64_t i = 0; i < count; ++i) {
        out[static_cast<std::size_t>(i)] = {ts[static_cast<std::size_t>(i)], close[static_cast<std::size_t>(i)]};
    }
    return out;
}

// Faithful reference port of CalculateMeanReversionSpeed (StudyHelperFunctions.cpp:3033),
// operating on a chronological in-memory price window instead of sc.BaseData. Returns
// both the clamped score and the SIGNED (current - median) deviation before abs(), needed
// here to classify forward-return direction (the production function only returns the
// abs-valued, elasticity-gated score).
//
// FIXED 2026-09-04: this port was stale, still using mean/std -- CalculateMeanReversionSpeed
// was reformulated to median/MAD (Kim & White 2004) 2026-08-31/committed 2026-09-02, but this
// tool was never updated to match, silently invalidating its own "empirically null" verdict
// (that result was measuring the OLD, already-abandoned formula). Re-ported verbatim against
// the current real function before re-running.
struct TimeBarResult { float score; float signedDeviation; };

TimeBarResult TimeBarMeanRevZ(const std::vector<float>& prices, int endIdxInclusive, int n) {
    n = std::clamp(n, 5, 40);
    const int start_idx = endIdxInclusive - n + 1;
    if (start_idx < 0) return {0.0f, 0.0f};

    constexpr double kPriceEps = 1e-6;
    constexpr double kMadConsistency = 1.4826;
    std::array<double, 40> log_prices{};
    for (int i = 0; i < n; ++i) {
        const double p = std::max(static_cast<double>(prices[static_cast<std::size_t>(start_idx + i)]), kPriceEps);
        log_prices[static_cast<std::size_t>(i)] = std::log(p);
    }

    std::array<double, 40> scratch{};
    std::copy_n(log_prices.begin(), n, scratch.begin());
    const int priceMid = n / 2;
    std::nth_element(scratch.begin(), scratch.begin() + priceMid, scratch.begin() + n);
    const double median_log_p = scratch[static_cast<std::size_t>(priceMid)];

    for (int i = 0; i < n; ++i) scratch[static_cast<std::size_t>(i)] = std::fabs(log_prices[static_cast<std::size_t>(i)] - median_log_p);
    std::nth_element(scratch.begin(), scratch.begin() + priceMid, scratch.begin() + n);
    const double mad_log_p = scratch[static_cast<std::size_t>(priceMid)];
    const double scale_log_p = mad_log_p * kMadConsistency;
    if (scale_log_p < 1e-6) return {0.0f, 0.0f};

    const double current_log_p = std::log(std::max(static_cast<double>(prices[static_cast<std::size_t>(endIdxInclusive)]), kPriceEps));
    const double signedDeviation = (current_log_p - median_log_p) / scale_log_p;
    const double abs_z_price = std::fabs(signedDeviation);

    const int m = n - 1;
    if (m < 3) {
        return {std::clamp(static_cast<float>(abs_z_price), 0.0f, 5.0f), static_cast<float>(signedDeviation)};
    }

    std::array<double, 40> returns{};
    for (int i = 0; i < m; ++i) {
        const int idx = start_idx + i + 1;
        const double p = std::max(static_cast<double>(prices[static_cast<std::size_t>(idx)]), kPriceEps);
        const double p_prev = std::max(static_cast<double>(prices[static_cast<std::size_t>(idx - 1)]), kPriceEps);
        returns[static_cast<std::size_t>(i)] = std::log(p / p_prev);
    }
    std::array<double, 40> returnScratch{};
    std::copy_n(returns.begin(), m, returnScratch.begin());
    const int retMid = m / 2;
    std::nth_element(returnScratch.begin(), returnScratch.begin() + retMid, returnScratch.begin() + m);
    const double medianR = returnScratch[static_cast<std::size_t>(retMid)];

    double num = 0.0, den = 0.0;
    for (int t = 1; t < m; ++t) {
        const double r_t = returns[static_cast<std::size_t>(t)] - medianR;
        const double r_prev = returns[static_cast<std::size_t>(t - 1)] - medianR;
        num += r_t * r_prev;
        den += r_prev * r_prev;
    }
    const double rho = (den > 1e-12) ? (num / den) : 0.0;
    const double elasticity_gate = std::clamp(1.0 - std::max(rho, 0.0), 0.0, 1.0);
    const double score = abs_z_price * elasticity_gate;
    return {std::clamp(static_cast<float>(score), 0.0f, 5.0f), static_cast<float>(signedDeviation)};
}

}  // namespace

int main(int argc, char** argv) {
    if (argc != 4 && argc != 5) {
        std::cerr << "usage: " << argv[0] << " <ticks_1s.bin> <bars_15m.bin> <signals_out.csv> [imbalance_threshold]\n";
        return 1;
    }
    const float imbalanceThreshold = (argc == 5) ? static_cast<float>(std::atof(argv[4])) : kDefaultImbalanceThreshold;
    const auto ticks = LoadTicks(argv[1]);
    const auto bars15m = LoadBars15m(argv[2]);
    std::cerr << "ticks (1s bars): " << ticks.size() << ", 15m bars: " << bars15m.size() << "\n";

    std::ofstream out(argv[3]);
    out << "timestamp_us,price,score,signed_deviation,variant\n";

    // --- Time-bar variant (15-min, CalculateMeanReversionSpeed's own math) ---
    // Uses n=20 (fixed representative window, mid of the [10,40] adaptive
    // observation_window_n range this dim actually runs on in production --
    // a single fixed value here since the adaptive-window derivation itself
    // is orthogonal to this specific measurement).
    {
        std::vector<float> prices(bars15m.size());
        for (std::size_t i = 0; i < bars15m.size(); ++i) prices[i] = bars15m[i].close;
        std::size_t signalCount = 0;
        for (int i = 20; i < static_cast<int>(bars15m.size()); ++i) {
            const auto r = TimeBarMeanRevZ(prices, i, 20);
            if (r.score > kGateThreshold) {
                out << bars15m[static_cast<std::size_t>(i)].timestamp_us << ","
                    << bars15m[static_cast<std::size_t>(i)].close << ","
                    << r.score << "," << r.signedDeviation << ",time_bar\n";
                ++signalCount;
            }
        }
        std::cerr << "time_bar signals (score > " << kGateThreshold << "): " << signalCount << "\n";
    }

    // --- Activity-clock variant (imbalance bars from 1s aggregate ticks) ---
    {
        ImbalanceBarEngine engine;
        engine.SetImbalanceThresholdForTesting(imbalanceThreshold);
        std::deque<float> returnBuffer;  // rolling last-100 returns, chronological (oldest first)
        std::size_t completedBefore = 0;
        std::size_t signalCount = 0;

        for (int i = 0; i < static_cast<int>(ticks.size()); ++i) {
            engine.OnTickWithPrice(/*barIndex=*/i, ticks[static_cast<std::size_t>(i)].ask_volume,
                                    ticks[static_cast<std::size_t>(i)].bid_volume,
                                    ticks[static_cast<std::size_t>(i)].close);
            const std::size_t completedNow = engine.GetCompletedBarCount();
            if (completedNow != completedBefore) {
                // A new imbalance bar just closed. Fetch its return (the most recent
                // element of the engine's own buffer) and append to our rolling window.
                float rawReturns[ImbalanceBarEngine::kImbalanceBarBufferCapacity];
                const std::size_t n = engine.GetImbalanceBarReturns(1, rawReturns);
                if (n == 1) {
                    returnBuffer.push_back(rawReturns[0]);
                    if (returnBuffer.size() > kReturnBufferN) returnBuffer.pop_front();
                }
                completedBefore = completedNow;

                if (returnBuffer.size() == kReturnBufferN) {
                    std::vector<float> window(returnBuffer.begin(), returnBuffer.end());
                    const float score = ActivityClockMeanRevZ(window.data(), kReturnBufferN);
                    if (std::isfinite(score) && score > kGateThreshold) {
                        // Recompute the signed deviation the same way ActivityClockMeanRevZ
                        // does internally (cumulative-path MEDIAN/MAD z-score, Kim & White 2004
                        // -- FIXED 2026-09-04, this previously used a stale mean/std
                        // recomputation even after the header itself was reformulated
                        // 2026-09-02), since the header only returns the final abs-valued/
                        // elasticity-gated score.
                        constexpr double kMadConsistency = 1.4826;
                        std::array<double, kReturnBufferN + 1> path{};
                        double cumsum = 0.0;
                        path[0] = 0.0;
                        for (int k = 0; k < kReturnBufferN; ++k) {
                            cumsum += static_cast<double>(window[static_cast<std::size_t>(k)]);
                            path[static_cast<std::size_t>(k + 1)] = cumsum;
                        }
                        constexpr int count = kReturnBufferN + 1;
                        constexpr int pathMid = count / 2;
                        std::array<double, count> scratch = path;
                        std::nth_element(scratch.begin(), scratch.begin() + pathMid, scratch.begin() + count);
                        const double medianPath = scratch[static_cast<std::size_t>(pathMid)];
                        for (int k = 0; k < count; ++k) scratch[static_cast<std::size_t>(k)] = std::fabs(path[static_cast<std::size_t>(k)] - medianPath);
                        std::nth_element(scratch.begin(), scratch.begin() + pathMid, scratch.begin() + count);
                        const double madPath = scratch[static_cast<std::size_t>(pathMid)];
                        const double scalePath = madPath * kMadConsistency;
                        const double signedDeviation = (scalePath > 1e-9) ? (cumsum - medianPath) / scalePath : 0.0;

                        out << ticks[static_cast<std::size_t>(i)].timestamp_us << ","
                            << ticks[static_cast<std::size_t>(i)].close << ","
                            << score << "," << signedDeviation << ",activity_clock\n";
                        ++signalCount;
                    }
                }
            }
        }
        std::cerr << "activity_clock imbalance bars formed: " << completedBefore << "\n";
        std::cerr << "activity_clock signals (score > " << kGateThreshold << "): " << signalCount << "\n";
    }

    return 0;
}

File: tools/observation_vector/mean_rev_z_variant_comparison.py
#!/usr/bin/env python3
"""mean_rev_z_variant_comparison.py -- empirical additive-vs-replace decision
for mean_rev_z/fast_mean_rev_z (docs/superpowers/plans/2026-08-28-activity-
clock-mean-rev-hurst-recurrence.md Task 5, per the sibling's revised verdict
in SCRATCHPAD.md: correlation alone does not decide this -- feed each
variant through Scoring.cpp:305's actual gate condition (score > 2.0) and
measure forward-return/hit-rate on the resulting signals independently).

Uses polars (columnar, lazy) for parquet I/O and numpy for the C++ driver's
raw binary export -- NOT pandas/.to_csv() (a 38.5M-row CSV round-trip is the
wrong tool for this data volume; contiguous binary arrays are what the C++
driver actually wants to fread(), not text it has to parse).

Pipeline:
1. Export real MES 1-second ticks (mes_continuous_ticks.parquet) and 15-min
   bars (mes_ripple_15m.parquet) to raw binary for the C++ driver
   (tools/observation_vector/mean_rev_z_variant_comparison.cpp, which uses the REAL
   ActivityClockMeanRevZ header + a faithful CalculateMeanReversionSpeed
   port to generate gate-crossing signals for both variants).
2. For each signal, look up the price `horizon_minutes` later in the full
   1-second series and compute the forward log-return.
3. Classify a "hit" as forward return having the OPPOSITE sign of the
   signal's own signed deviation (price was stretched one way, hit = it
   moved back the other way -- the literal definition of "reversion").
4. Report N signals, hit rate, mean/median forward return, per variant --
   plus the raw correlation between the two score series as a secondary
   diagnostic (not the deciding test, per the sibling's correction).

Usage: mamba run -n mts python tools/observation_vector/mean_rev_z_variant_comparison.py \
           [--horizon-minutes 60] [--imbalance-threshold 15.0] [--calibrate-only]
"""

import argparse
import math
import struct
import subprocess
import tempfile
from pathlib import Path

import numpy as np
import polars as pl


WORKSPACE = Path("/home/rcruz/devel/VSCode")
TICKS_PARQUET = WORKSPACE / "lbrnet/data/raw/mes_continuous_ticks.parquet"
BARS_15M_PARQUET = WORKSPACE / "lbrnet/data/raw/mes_ripple_15m.parquet"
MINDFULTRADER = WORKSPACE / "MindfulTrader"
DRIVER_SRC = MINDFULTRADER / "tools/observation_vector/mean_rev_z_variant_comparison.cpp"
DRIVER_BIN = MINDFULTRADER / "tools/bin/mean_rev_z_variant_comparison"


def build_driver():
    DRIVER_BIN.parent.mkdir(parents=True, exist_ok=True)
    print(
        f"building driver: g++ -O2 -std=c++17 -I{MINDFULTRADER / 'include'} {DRIVER_SRC} -o {DRIVER_BIN}"
    )
    subprocess.run(
        [
            "g++",
            "-O2",
            "-std=c++17",
            "-I",
            str(MINDFULTRADER / "include"),
            str(DRIVER_SRC),
            "-o",
            str(DRIVER_BIN),
        ],
        check=True,
    )


def write_binary(path: Path, timestamp_us: np.ndarray, *float_cols: np.ndarray):
    """[int64 count][int64 timestamp_us x count][float32 col x count]..."""
    count = timestamp_us.shape[0]
    with open(path, "wb") as f:
        f.write(struct.pack("<q", count))
        f.write(timestamp_us.astype("<i8", copy=False).tobytes())
        for col in float_cols:
            f.write(col.astype("<f4", copy=False).tobytes())


def export_binaries(scratch_dir: Path):
    ticks = (
        pl.scan_parquet(TICKS_PARQUET)
        .select("timestamp_us", "close", "bid_volume", "ask_volume")
        .sort("timestamp_us")
        .collect()
    )
    ticks_bin = scratch_dir / "ticks_1s.bin"
    write_binary(
        ticks_bin,
        ticks["timestamp_us"].to_numpy(),
        ticks["close"].to_numpy(),
        ticks["bid_volume"].to_numpy(),
        ticks["ask_volume"].to_numpy(),
    )
    print(f"exported {len(ticks)} 1-second bars to {ticks_bin}")

    bars = (
        pl.scan_parquet(BARS_15M_PARQUET)
        .select("timestamp_us", "close")
        .sort("timestamp_us")
        .collect()
    )
    bars_bin = scratch_dir / "bars_15m.bin"
    write_binary(bars_bin, bars["timestamp_us"].to_numpy(), bars["close"].to_numpy())
    print(f"exported {len(bars)} 15-min bars to {bars_bin}")

    return ticks, ticks_bin, bars_bin


def compute_forward_returns(
    signals: pl.DataFrame, ticks: pl.DataFrame, horizon_minutes: int
) -> pl.DataFrame:
    horizon_us = horizon_minutes * 60 * 1_000_000
    ts = ticks["timestamp_us"].to_numpy()
    px = ticks["close"].to_numpy()

    sig_ts = signals["timestamp_us"].to_numpy()
    sig_price = signals["price"].to_numpy()
    target_ts = sig_ts + horizon_us
    idx = np.searchsorted(ts, target_ts)

    valid = idx < len(ts)
    # Guard against a huge gap (overnight/weekend) swamping the horizon.
    gap_ok = np.zeros_like(valid)
    gap_ok[valid] = (ts[idx[valid]] - sig_ts[valid]) <= horizon_us * 3
    valid &= gap_ok

    forward_return = np.full(len(signals), np.nan)
    forward_return[valid] = np.log(px[idx[valid]] / sig_price[valid])

    return signals.with_columns(pl.Series("forward_return", forward_return)).filter(
        pl.col("forward_return").is_not_nan()
    )


def wilson_ci(k: int, n: int, z: float = 1.96) -> tuple:
    """Wilson score interval -- correct coverage near p=0.5 even at these n, unlike a naive p +/- 1.96*se."""
    phat = k / n
    denom = 1 + z * z / n
    center = (phat + z * z / (2 * n)) / denom
    half = (z / denom) * np.sqrt(phat * (1 - phat) / n + z * z / (4 * n * n))
    return center - half, center + half


def report(signals: pl.DataFrame, variant: str):
    sub = signals.filter(pl.col("variant") == variant)
    if len(sub) == 0:
        print(f"  {variant}: 0 signals with a valid forward return")
        return
    fwd = sub["forward_return"].to_numpy()
    dev = sub["signed_deviation"].to_numpy()
    # "Hit" = forward return moved opposite to the signal's own stretch direction.
    hits = np.sign(fwd) == -np.sign(dev)
    n = len(hits)
    k = int(hits.sum())
    hit_rate = k / n
    # Two-sided z-test against the null hit_rate = 0.5 (pure chance).
    se_null = np.sqrt(0.25 / n)
    z = (hit_rate - 0.5) / se_null
    p_value = float(2 * (1 - 0.5 * (1 + math.erf(abs(z) / math.sqrt(2)))))
    lo, hi = wilson_ci(k, n)
    print(
        f"  {variant}: n={n}  hit_rate={hit_rate:.4f}  95%CI=[{lo:.4f}, {hi:.4f}]  "
        f"z={z:+.2f}  p={p_value:.3f} (vs null=0.5)  "
        f"mean_fwd_ret={np.mean(fwd):.6f}  median_fwd_ret={np.median(fwd):.6f}  "
        f"std_fwd_ret={np.std(fwd):.6f}"
    )


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--horizon-minutes",
        type=int,
        default=60,
        help="Forward-return horizon in minutes, applied identically to both variants",
    )
    parser.add_argument(
        "--imbalance-threshold",
        type=float,
        default=None,
        help="Override the activity-clock imbalance threshold (calibrate bar-formation rate first)",
    )
    parser.add_argument(
        "--calibrate-only",
        action="store_true",
        help="Just run the driver and print bar-formation/signal counts, skip forward-return stats",
    )
    parser.add_argument(
        "--reuse-binaries",
        type=str,
        default=None,
        help="Path to a scratch dir with already-exported ticks_1s.bin/bars_15m.bin (skip re-export)",
    )
    args = parser.parse_args()

    build_driver()

    scratch_ctx = tempfile.TemporaryDirectory(prefix="mean_rev_z_comparison_")
    scratch_dir = Path(args.reuse_binaries) if args.reuse_binaries else Path(scratch_ctx.name)

    if args.reuse_binaries and (scratch_dir / "ticks_1s.bin").exists():
        print(f"reusing existing binaries in {scratch_dir}")
        ticks = (
            pl.scan_parquet(TICKS_PARQUET)
            .select("timestamp_us", "close")
            .sort("timestamp_us")
            .collect()
        )
        ticks_bin = scratch_dir / "ticks_1s.bin"
        bars_bin = scratch_dir / "bars_15m.bin"
    else:
        scratch_dir.mkdir(parents=True, exist_ok=True)
        ticks, ticks_bin, bars_bin = export_binaries(scratch_dir)

    signals_csv = scratch_dir / "signals.csv"
    cmd = [str(DRIVER_BIN), str(ticks_bin), str(bars_bin), str(signals_csv)]
    if args.imbalance_threshold is not None:
        cmd.append(str(args.imbalance_threshold))
    print(f"running driver: {' '.join(cmd)}")
    subprocess.run(cmd, check=True)

    if args.calibrate_only:
        if not args.reuse_binaries:
            scratch_ctx.cleanup()
        return

    signals = pl.read_csv(signals_csv)
    print(f"\ntotal signals (both variants, before forward-return filtering): {len(signals)}")

    signals = compute_forward_returns(signals, ticks, args.horizon_minutes)
    print(f"signals with a valid {args.horizon_minutes}-min forward return: {len(signals)}\n")

    print(f"=== Forward-return / hit-rate report (horizon={args.horizon_minutes}min) ===")
    report(signals, "time_bar")
    report(signals, "activity_clock")

    print(
        "\n[note] no same-timestamp score correlation computed here -- the two variants fire "
        "signals on different native clocks (15-min bars vs imbalance bars), so a naive "
        "paired correlation would require resampling one onto the other's grid, which this "
        "script deliberately does not do implicitly. The forward-return/hit-rate comparison "
        "above is the decision-relevant test per the sibling's verdict."
    )

    if not args.reuse_binaries:
        scratch_ctx.cleanup()


if __name__ == "__main__":
    main()
